### ISSUE TITLE
magit-git-executable default should be updated for Git for Windows 2.x

### ISSUE THREAD

**Issue by npostavs** [magit/magit#2259]
I'm opening this issue mainly to have a non google groups place for the following important bit of info.

[Johannes Schindelin wrote](https://groups.google.com/d/msg/git-for-windows/RYqgFyJLKwM/2uMMOjoTDQAJ):

> The first thing to note is that I made a conscious decision with Git for
Windows 1.x to just slam every executable into /bin/ and be done with
it. This breaks the paradigm of MinGW, though, which makes a very clear
distinction between MinGW executables (i.e. programs using the pure
Win32 API) and MSys executables (i.e. programs relying on the POSIX
emulation of MSys). The former are expected to be in /mingw/bin/ and the
latter in /bin/ for MSys1 (on which Git for Windows 1.x is based).

> This was a mistake and the major version jump to 2.x gave me an
excellent opportunity to fix it. Now it is even more complicated,
though: there are *three* bin/ directories: /usr/bin/ for MSys2
executables, /mingw32/bin/ and /mingw64/bin/ for 32-bit and 64-bit
executables, respectively, that do not rely on the POSIX emulation. It
is not possible to mix and match 32-bit and 64-bit MSys2 programs,
though.

> For 32-bit Git for Windows, I opted to stay with an all 32-bit
environment, so all the MinGW executables are in /mingw32/bin/. For
64-bit, consequently, they live in /mingw64/bin/.

> Now, for backwards-compatibility, I introduced what I call "redirectors"
into the /bin/ directory of Git for Windows 2.x. There is `git.exe`,
`bash.exe` and `sh.exe` in this directory, and nothing else. And those
executables are not the "real things", either: they just execute the
real executables in the correct places, wait for them to exit and then
exit with the same exit status. 

Additional notes:

The `git.exe` "redirector" is still present in /cmd/, like in 1.x.

**Comment by npostavs** [comment]
Actually, the upshot might be that we shouldn't do anything. `/bin/git.exe` and `/cmd/git.exe` are now both (identical) wrappers, the real git lives in `/mingw32/bin/` but it needs to have `/usr/bin/` in `PATH` in order to find `ssh.exe`, and `ssh.exe` needs `/mingw32/bin/` in `PATH` to find some dlls.

The `/cmd/git.exe` wrapper does the necessary magic with the PATH (see [git-wrapper.c][1]), including selecting between mingw32 and mingw64 which seems a bit impractical to do from Emacs. If we leave things as they are, it should still work (albeit slightly slower than it could have been).

On the other hand, I'll probably need to work out how to perform the PATH magic for [ssh-agency][3] anyway...

---

Updating the benchmark from [before][2], I see basically the same timings.

```elisp
(require 'benchmark)

(dolist (git-dir '("c:/Program Files/Git/cmd"
                   "c:/Program Files/Git/bin"
                   "c:/Program Files/Git/mingw32/bin"))
  (let ((the-git (expand-file-name "git.exe" git-dir)))
    (message "git = %s" the-git)
    (benchmark 100 '(call-process the-git nil nil nil "status"))))
```

```
git = c:/Program Files/Git/cmd/git.exe
Elapsed time: 3.994000s (0.000000s in 1 GCs)
git = c:/Program Files/Git/bin/git.exe
Elapsed time: 3.962000s
git = c:/Program Files/Git/mingw32/bin/git.exe
Elapsed time: 3.058000s (0.016000s in 1 GCs)
```

[1]: https://github.com/git-for-windows/git/blob/master/compat/win32/git-wrapper.c
[2]: https://github.com/magit/magit/pull/1289
[3]: https://github.com/magit/ssh-agency

**Comment by tarsius** [comment]
Being 1/4 faster isn't such a small improvement, so if you can figure out how to use "the real thing" then I would want to install that change.

**Comment by npostavs** [comment]
The "ask git" strategy from #2281 looks to the best way of doing this:
* `cmd/git -c "alias.exe=!which git | cygpath -wf -" exe` returns the "real" git (it actually returns the one in `libexec/git-core`, though that turns out to be byte identical to `mingw32/bin/git.exe` anyway).
* `cmd/git -c "alias.path=!cygpath -wp \"$PATH\"" path` returns the needed `PATH` value to run it

This returns reasonable answers for msys2 git as well, still need to check what happens with cygwin and "Git for Windows" 1.x

**Comment by npostavs** [comment]
@tarsius I've pushed [/np/win32-git2.x] which I think should work. I'm not sure how to set both the initial value for `magit-git-executable` and the extra PATH setting in a nice way though.

[/np/win32-git2.x]: https://github.com/magit/magit/compare/np/win32-git2.x

**Comment by tarsius** [comment]
Does the new approach only work with cygwin? What about someone using just gfw 2, would that still cause the "wrong" executable to be use? (I think that would be the most common case, so we should do something about that.) Does `cygpath -wp \"$PATH\"` return the the complete path or just where cygwin can be found?

**Comment by npostavs** [comment]
> Does the new approach only work with cygwin? What about someone using just gfw 2, would that still cause the "wrong" executable to be use?

No, both MSys2 and gfw 2 also come with cygpath. It's only gfw 1 that lacks it.

The new approach does have the small downside of choosing the libexec/git-core executable which fails without extra PATH elements in some cases. But as far as I can tell, it does work just fine with the extra PATH elements and since we need code to add the extra path elements for ssh remotes with gfw 2 anyway, this doesn't cause any trouble in practice.

> Does `cygpath -wp \"$PATH\"` return the the complete path or just where cygwin can be found?

The complete path. We could chop off the original `PATH` value ourselves to get just the git-relevant part.

Below is some example output (`%ERRORLEVEL%` is the cmd.exe version of `$?`). You'll notice that the ssh host key is not recognized for Cygwin and Msys2. That is because I installed them only for testing and I haven't bothered to configure them to use my normal home directory.

#### Summary

Doing the equivalent of
```sh
git_exe=$(git -c "alias.exe=!which git | cygpath -wf -" exe)
git_path=$(git -c "alias.path=!cygpath -wp \"$PATH\"" path)
PATH=$git_path "$git_exe" <git-command>
```

works everywhere except Git for Windows 1.x (and cygwin when its bin is not on `PATH`). These cases can be detected easily because the probing commands fail.

#### Git for Windows 2.x portable
```
> c:\Git-2.x\cmd\git.exe --version
git version 2.5.3.windows.1

> c:\Git-2.x\cmd\git.exe -c "alias.exe=!which git | cygpath -wf -" exe
C:\Git-2.x\mingw64\libexec\git-core\git

> c:\Git-2.x\cmd\git.exe -c "alias.path=!cygpath -wp \"$PATH\"" path
C:\Git-2.x\mingw64\libexec\git-core;C:\Git-2.x\mingw64\bin;C:\Git-2.x\usr\bin;[...];C:\Users\npostavs\bin;C:\Program Files\apache-ant-1.9.4\bin

> C:\Git-2.x\mingw64\libexec\git-core\git --version
git version 2.5.3.windows.1

> C:\Git-2.x\mingw64\libexec\git-core\git fetch
error: cannot spawn ssh: No such file or directory
fatal: unable to fork
> set PATH=C:\Git-2.x\mingw64\libexec\git-core;C:\Git-2.x\mingw64\bin;C:\Git-2.x\usr\bin

> C:\Git-2.x\mingw64\libexec\git-core\git fetch
Enter passphrase for key '/c/Users/npostavs/.ssh/id_rsa':
... [successful fetch]
```

#### Git for Windows 2.x
```
> "c:\Program Files\Git\cmd\git.exe" --version
git version 2.5.0.windows.1

> "c:\Program Files\Git\cmd\git.exe" -c "alias.exe=!which git | cygpath -wf -" exe
C:\Program Files\Git\mingw32\libexec\git-core\git

> "c:\Program Files\Git\cmd\git.exe" -c "alias.path=!cygpath -wp \"$PATH\"" path
C:\Program Files\Git\mingw32\libexec\git-core;C:\Program Files\Git\mingw32\bin;C
:\Program Files\Git\usr\bin;...;C:\Users\npostavs\bin;C:\Program Files\Git\cmd

> "C:\Program Files\Git\mingw32\libexec\git-core\git" --version
Popup dialog : [The program can't start because libiconv-2.dll is missing from your computer. Try reinstall the program to fix this problem]
> echo %ERRORLEVEL%
-1073741515

> "c:\Program Files\Git\mingw32\bin\git.exe" --version
git version 2.5.0.windows.1

> "c:\Program Files\Git\mingw32\bin\git.exe" fetch
error: cannot spawn ssh: No such file or directory
fatal: unable to fork

> set PATH=C:\Program Files\Git\mingw32\libexec\git-core;C:\Program Files\Git\mingw32\bin;C:\Program Files\Git\usr\bin
> "C:\Program Files\Git\mingw32\libexec\git-core\git" fetch
Enter passphrase for key '/c/Users/npostavs/.ssh/id_rsa':
[successful fetch]
```

#### Git for Windows 1.x
```
> "c:\Program Files (x86)\Git\cmd\git.exe" --version
git version 1.9.5.msysgit.1

> "c:\Program Files (x86)\Git\cmd\git.exe" -c "alias.exe=!which git | cygpath -wf -" exe
which git | cygpath -wf -: cygpath: command not found
fatal: While expanding alias 'exe': 'which git | cygpath -wf -': No such file or
 directory

> echo %ERRORLEVEL%
128
```

#### Cygwin when `PATH` doesn't include cygwin's `bin`
```
> c:\cygwin\bin\git.exe --version
git version 2.5.1

> c:\cygwin\bin\git.exe -c "alias.exe=!which git | cygpath -wf -" exe
which git | cygpath -wf -: which: command not found
which git | cygpath -wf -: cygpath: command not found
fatal: While expanding alias 'exe': 'which git | cygpath -wf -': No such file or
 directory

> echo %ERRORLEVEL%
128
```

#### Cygwin when `PATH` includes cygwin's `bin`
```
> git --version
git version 2.5.1

> c:\cygwin\bin\git.exe -c "alias.exe=!which git | cygpath -wf -" exe
C:\cygwin\usr\libexec\git-core\git

> c:\cygwin\bin\git.exe -c "alias.path=!cygpath -wp \"$PATH\"" path
C:\cygwin\usr\libexec\git-core;C:\cygwin\bin;C:\Windows\system32;C:\Windows;[...]

> C:\cygwin\usr\libexec\git-core\git fetch
The authenticity of host 'github.com (192.30.252.129)' can't be established.
RSA key fingerprint is 16:27:ac:a5:76:28:2d:36:63:1b:56:4d:eb:df:a6:48.
```

#### MSys 2
```
> c:\msys64\usr\bin\git --version
git version 2.4.1

> c:\msys64\usr\bin\git -c "alias.exe=!which git | cygpath -wf -" exe
C:\msys64\usr\lib\git-core\git

> c:\msys64\usr\bin\git -c "alias.path=!cygpath -wp \"$PATH\"" path
C:\msys64\usr\lib\git-core;C:\msys64\usr\bin;...;C:\Users\npostavs\bin;C:\Program F
iles\Git\cmd

> C:\msys64\usr\lib\git-core\git --version
Popup dialog: [The program can't start because msys-crypto-1.0.0.dll is missing from your computer...]
> echo %ERRORLEVEL%
-1073741515
> c:\msys64\usr\bin\git fetch
Could not create directory '/home/npostavs/.ssh'.
The authenticity of host 'github.com (192.30.252.128)' can't be established.
RSA key fingerprint is 16:27:ac:a5:76:28:2d:36:63:1b:56:4d:eb:df:a6:48.

> set PATH=C:\msys64\usr\lib\git-core;C:\msys64\usr\bin
> C:\msys64\usr\lib\git-core\git fetch
Could not create directory '/home/npostavs/.ssh'.
The authenticity of host 'github.com (192.30.252.129)' can't be established.
RSA key fingerprint is 16:27:ac:a5:76:28:2d:36:63:1b:56:4d:eb:df:a6:48.
```

**Comment by tarsius** [comment]
> The new approach does have the small downside of choosing the libexec/git-core executable which fails without extra PATH elements in some cases. But as far as I can tell, it does work just fine with the extra PATH elements and since we need code to add the extra path elements for ssh remotes with gfw 2 anyway, this doesn't cause any trouble in practice.

Okay with me.

> > Does `cygpath -wp \"$PATH\"` return the the complete path or just where cygwin can be found?

The complete path. We could chop off the original `PATH` value ourselves to get just the git-relevant part.

I had the opposite in mind: if this did *not* return the complete path, then we would have to add "the rest" manually (for the sake of git hooks and such). But since it returns the complete value we are set.

**Comment by tarsius** [comment]
We should wait until after we know whether my fix for #2284 works, but once that has been taken care of, you can merge this when you feel it is ready.

**Comment by tarsius** [comment]
Maybe you should merge this now, instead of waiting until the obscure cases in #2284 are fully understood.
